src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkBatchInsertRecord.java

```diff
@@ -0,0 +1,319 @@
+                // Formatter to remove the decimal part as SQL Server floors the decimal in integer types
+                DecimalFormat decimalFormatter = new DecimalFormat("#");
+                decimalFormatter.setRoundingMode(RoundingMode.DOWN);
+                String formatedfInput = decimalFormatter.format(Double.parseDouble(data.toString()));
+                return Integer.valueOf(formatedfInput);
+            }
+
+            case Types.TINYINT:
+            case Types.SMALLINT: {
+                // Formatter to remove the decimal part as SQL Server floors the decimal in integer types
+                DecimalFormat decimalFormatter = new DecimalFormat("#");
+                decimalFormatter.setRoundingMode(RoundingMode.DOWN);
+                String formatedfInput = decimalFormatter.format(Double.parseDouble(data.toString()));
+                return Short.valueOf(formatedfInput);
+            }
+
+            case Types.BIGINT: {
+                BigDecimal bd = new BigDecimal(data.toString().trim());
+                try {
+                    return bd.setScale(0, RoundingMode.DOWN).longValueExact();
+                }
+                catch (ArithmeticException ex) {
+                    String value = "'" + data + "'";
+                    MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_errorConvertingValue"));
+                    throw new SQLServerException(form.format(new Object[] {value, JDBCType.of(cm.columnType)}), null, 0, ex);
+                }
+            }
+
+            case Types.DECIMAL:
+            case Types.NUMERIC: {
+                BigDecimal bd = new BigDecimal(data.toString().trim());
+                return bd.setScale(cm.scale, RoundingMode.HALF_UP);
+            }
+
+            case Types.BIT: {
+                // "true" => 1, "false" => 0
+                // Any non-zero value (integer/double) => 1, 0/0.0 => 0
+                try {
+                    return (0 == Double.parseDouble(data.toString())) ? Boolean.FALSE : Boolean.TRUE;
+                }
+                catch (NumberFormatException e) {
+                    return Boolean.parseBoolean(data.toString());
+                }
+            }
+
+            case Types.REAL: {
+                return Float.parseFloat(data.toString());
+            }
+
+            case Types.DOUBLE: {
+                return Double.parseDouble(data.toString());
+            }
+
+            case Types.BINARY:
+            case Types.VARBINARY:
+            case Types.LONGVARBINARY:
+            case Types.BLOB: {
+                // Strip off 0x if present.
+                String binData = data.toString().trim();
+                if (binData.startsWith("0x") || binData.startsWith("0X")) {
+                    return binData.substring(2);
+                }
+                else {
+                    return binData;
+                }
+            }
+
+            case 2013:    // java.sql.Types.TIME_WITH_TIMEZONE
```

Comment:
why do we have to use hardcoded 2013 instead of Types. TIME_WITH_TIMEZONE here? same with 2014 and TIMESTAMP_WITH_TIMEZONE?

Reply:
Looking at the code for java.sql.Types, looks like TIME_WITH_TIMEZONE and TIMESTAMP_WITH_TIMEZONE was added later on in JDBC 4.2. They probably wrote this code before those values were added to the Types file, but knew the number code beforehand. I've made the changes to use the actual name.